sky/serve/controller.py

```diff
@@ -85,6 +87,37 @@ def _run_autoscaler(self):
                     logger.error(f'  Traceback: {traceback.format_exc()}')
             time.sleep(self._autoscaler.get_decision_interval())
 
+    def _purge_replica(
+            self, replica_info: replica_managers.ReplicaInfo) -> Dict[str, str]:
+        replica_id = replica_info.replica_id
+
+        logger.info(f'Purging replica {replica_id}...')
+
+        if replica_info.status not in serve_state.ReplicaStatus.failed_statuses(
+        ):
+            return {
+                'message': f'No purging for replica {replica_id} since '
+                           f'the replica does not have a failed status.'
+            }
```

Comment:
Should we allow `--purge` to terminate a healthy replica as well, to align the semantic of `sky serve down`?

Reply:
Good point. I thought "purge" means cleaning failed instances in the project.

Reply:
Actually, the current behavior of `sky serve down` already aligns with this semantic. The `--purge` option doesn't affect healthy replicas; it's specifically designed to force termination of failed services/replicas. Healthy ones are terminated normally regardless of the `--purge` flag.

In sky/serve/serve_utils.py#L295-L320:

```python
if (service_status is None or service_status['status']
        in serve_state.ServiceStatus.failed_statuses()):
    if purge:
        message = _terminate_failed_services(service_name, failed_status)
    else:
        messages.append(
            f'{colorama.Fore.YELLOW}Service {service_name!r} is in '
            f'failed status ({failed_status}). Skipping '
            'its termination as it could lead to a resource leak. '
            f'(Use `sky serve down {service_name} --purge` to '
            'forcefully terminate the service.)'
            f'{colorama.Style.RESET_ALL}')
else:
    # Send the terminate signal to controller.
    signal_file = pathlib.Path(
        constants.SIGNAL_FILE_PATH.format(service_name))
    with filelock.FileLock(str(signal_file) + '.lock'):
        with signal_file.open(mode='w', encoding='utf-8') as f:
            f.write(UserSignal.TERMINATE.value)
            f.flush()
```

Reply:
I think the current implementation of `sky serve down --purge` can tear down healthy service as well, since we are writing the terminate signal regardless of purge or not (in the `else` branch). However, iiuc `sky serve down --purge --replica-id 1` cannot terminate replica 1 if replica 1 is not in a failed status?

Reply:
I see.

Reply:
Fixed [08961c](https://github.com/skypilot-org/skypilot/pull/4032/commits/08961c8918dd6433204231770e9aa618a74694b6)